pkg/timeutils/utils.go

```diff
@@ -1,36 +1,122 @@
 package timeutils
 
 import (
+	"fmt"
+	"math"
 	"strconv"
 	"strings"
 	"time"
 )
 
+// These are additional predefined layouts for use in Time.Format and Time.Parse
+// with --since and --until parameters for `docker logs` and `docker events`
+const (
+	RFC3339Local     = "2006-01-02T15:04:05"           // RFC3339 with local timezone
+	RFC3339NanoLocal = "2006-01-02T15:04:05.999999999" // RFC3339Nano with local timezone
+	DateWithZone     = "2006-01-02Z07:00"              // RFC3339 with time at 00:00:00
+	DateLocal        = "2006-01-02"                    // RFC3339 with local timezone and time at 00:00:00
+)
+
 // GetTimestamp tries to parse given string as golang duration,
 // then RFC3339 time and finally as a Unix timestamp. If
 // any of these were successful, it returns a Unix timestamp
 // as string otherwise returns the given value back.
 // In case of duration input, the returned timestamp is computed
 // as the given reference time minus the amount of the duration.
-func GetTimestamp(value string, reference time.Time) string {
+func GetTimestamp(value string, reference time.Time) (string, error) {
 	if d, err := time.ParseDuration(value); value != "0" && err == nil {
-		return strconv.FormatInt(reference.Add(-d).Unix(), 10)
+		return strconv.FormatInt(reference.Add(-d).Unix(), 10), nil
 	}
 
 	var format string
+	var parseInLocation bool
+
+	// if the string has a Z or a + or three dashes use parse otherwise use parseinlocation
+	parseInLocation = !(strings.ContainsAny(value, "zZ+") || strings.Count(value, "-") == 3)
+
 	if strings.Contains(value, ".") {
-		format = time.RFC3339Nano
+		if parseInLocation {
+			format = RFC3339NanoLocal
+		} else {
+			format = time.RFC3339Nano
+		}
+	} else if strings.Contains(value, "T") {
+		// we want the number of colons in the T portion of the timestamp
+		tcolons := strings.Count(value, ":")
+		// if parseInLocation is off and we have a +/- zone offset (not Z) then
+		// there will be an extra colon in the input for the tz offset subract that
+		// colon from the tcolons count
+		if !parseInLocation && !strings.ContainsAny(value, "zZ") && tcolons > 0 {
+			tcolons--
+		}
+		if parseInLocation {
+			if tcolons >= 2 {
+				format = RFC3339Local
+			} else if tcolons == 1 {
+				format = "2006-01-02T15:04"
+			} else {
+				format = "2006-01-02T15"
+			}
+		} else {
+			if tcolons >= 2 {
+				format = time.RFC3339
+			} else if tcolons == 1 {
+				format = "2006-01-02T15:04Z07:00"
+			} else {
+				format = "2006-01-02T15Z07:00"
+			}
+		}
+	} else if parseInLocation {
```

Comment:
This looks like a good case for a `switch` instead of `if...else if...elise if..`